File: XML_PARSER.py
from USER import User, Post
from graph import Graph


def read_file(address):
    # address should be (r'absolute address') #
    with open(address, 'r') as file:
        initial_file = file.read()
        print(initial_file)
        return initial_file


def write_file(address, last_file):
    # address should be (r'absolute address') #
    with open(address, 'w') as file:
        file.write(last_file)


def parsing_with_correcting_errors(address):
    initial_file = read_file(address)
    end_file = ""
    line_num = 1
    expected = ['<users>', '<user>', '<id>', '<name>', '<posts>', '<post>', '<body>', '<topics>', '<topic>',
                '<followers>', '<follower>', '<id>']
    expected_index = 0
    last_opening = []
    next_is_opening = 1
    next_is_closing = 0

    idd = ""
    name = ""
    body = ""
    topics = []
    posts = []
    followers_ids = []
    first_id = 1
    g = Graph()
    index = 0
    while index < len(initial_file):

        c = initial_file[index]

        if c == ' ' or c == '\n':
            end_file = end_file + c
            if c == '\n':
                line_num += 1

            index += 1
            continue

        if next_is_opening or next_is_closing:
            if initial_file[index + 1] == '/':
                next_is_opening = 0
                next_is_closing = 1
            else:
                next_is_opening = 1
                next_is_closing = 0

        if next_is_opening and (not next_is_closing):
            last_opening.append(expected[expected_index])
            for ch in expected[expected_index]:
                end_file = end_file + ch
                index += 1

            # 1)
            # how to know error??
            # no_error = 1
            # for i in range(len(expected[expected_index])):
            # no_error = no_error and (expected[expected_index][i] == initial_file[index])
            # index += 1

            if expected[expected_index] == '<id>' or expected[expected_index] == '<name>' or expected[
                expected_index] == '<body>' or expected[expected_index] == '<topic>':
                next_is_opening = 0
                next_is_closing = 0

            if expected_index != 8 and expected_index != 11:
                expected_index += 1

            elif expected_index == 11:
                expected_index -= 1



        elif not next_is_opening and not next_is_closing:
            txt = ""
            while initial_file[index] != '<':
                end_file = end_file + initial_file[index]
                txt = txt + initial_file[index]
                if initial_file[index] == '\n':
                    line_num += 1

                index += 1

            if last_opening[-1] == '<id>':
                if first_id:
                    first_id = 0
                    idd = txt.strip()
                else:
                    followers_ids.append(txt)
            elif last_opening[-1] == '<name>':
                name = txt.strip()
            elif last_opening[-1] == '<body>':
                body = txt.strip()
            elif last_opening[-1] == '<topic>':
                topics.append(txt.strip())

            next_is_opening = 0
            next_is_closing = 1


        else:
            # 1)
            # error checking
            # how to??

            open_tag = last_opening[-1]
            close_tag = open_tag[0] + '/' + open_tag[1:]
            index += len(close_tag)
            last_opening.pop()
            for ch in close_tag:
                end_file = end_file + ch

            if close_tag == '</post>':
                post1 = Post(body, topics)
                posts.append(post1)
                body = ""
                topics = []

            elif close_tag == '</user>':
                user = User(idd, name, posts)
                g.add_user(idd, user)
                for x in followers_ids:
                    g.add_follower(idd, x)

                first_id = 1
                idd = ""
                name = ""
                posts = []
                followers_ids = []

            elif close_tag == '</topics>':
                expected_index = 5

            elif close_tag == '</posts>':
                expected_index = 9

            elif close_tag == '</followers>':
                expected_index = 1

            elif close_tag == '</users>':
                break

            next_is_opening = 1
            next_is_closing = 0

    return end_file, g


end_file, g = parsing_with_correcting_errors(r'E:\College\Data structures\Project\DS_Project\sample.xml')
print('\n****************************\n')
print(end_file)

File: XML_Parser.cpp
#include "XML_Parser.h"
void write_to_file(string address, string& content) {
    ofstream file(address);
    if (file.is_open()) {
        file << content;
        file.close();
    } else {
        cerr << "Error: Unable to write to file" << std::endl;
    }
}
string read_file(string address) {
    ifstream file(address);
    if (!file.is_open()) {
        cerr << "Error: Unable to open file" << std::endl;
        return "";
    }
    string initial_file;
    string line;
    while (getline(file, line)) {
        initial_file += line + '\n';
    }
    file.close();
    return initial_file;
}
bool temp_is_dummy(string &temp) {
    bool is_dummy = 1;
    for (auto ch: temp)
        is_dummy = is_dummy && (ch == '\n' || ch == ' ');


    return is_dummy;
}
string trim(string&x) {
    string y = "";
    int start = 0;
    for (char c: x) {
        if (c != '\n' && c != ' ')
            break;

        start++;
    }
    int end = x.size() - 1;
    for (int i = x.size() - 1; i >= 0; i--) {
        if (x[i] != '\n' && x[i] != ' ')
            break;

        end--;
    }

    for (int i = start; i <= end; i++) {
        y += x[i];
    }
    return y;
}
vector<string> divide_string_for_graph(string &file) {
    vector<string> divided_file;
    string temp = "";
    bool start = true;
    bool tag;
    int index = 0;
    while (index < file.size()) {
        if (start) {
            if (file[index] == '<')
                tag = true;
            else
                tag = false;

            start = false;
        }
        temp += file[index];
        if (tag) {
            if (file[index] == '>') {
                divided_file.push_back(temp);
                temp = "";
                start = true;
            }
        }
        else {
            if (file[index + 1] == '<') {
                if (!temp_is_dummy(temp)) {
                    divided_file.push_back(trim(temp));
                }
                temp = "";
                start = true;
            }
        }
        index++;
    }
    return divided_file;
}
vector<string> divide_string_for_correction(string &file) {
    vector<string> divided_file;
    string temp = "";
    bool start = true;
    bool tag;
    int index = 0;
    while (index < file.size()) {
        if (start) {
            if (file[index] == '<')
                tag = true;
            else
                tag = false;

            start = false;
        }
        temp += file[index];
        if (tag) {
            if (file[index] == '>') {
                divided_file.push_back(temp);
                temp = "";
                start = true;
            }
        } else {
            if (file[index + 1] == '<') {
                divided_file.push_back(temp);

                temp = "";
                start = true;
            }
        }
        index++;
    }
    return divided_file;
}
Graph parse(string &file) {
    Graph g;
    vector<string> divided_file = divide_string_for_graph(file);
    User* obj_ptr;
    Post* post_ptr;
    vector<string> topics;
    vector<Post> posts;
    bool inside_followers = false;
    for(int i=0;i<divided_file.size();i++){
        string s = divided_file[i];
        if(s=="<user>")
            obj_ptr = new User;


        if(s=="<id>") {
            if (!inside_followers)
                obj_ptr->set_user_id(stoi(divided_file[i + 1]));
            else {
                g.add_follower(obj_ptr->get_user_id(), stoi(divided_file[i + 1]));
            }
        }

        if(s=="<name>")
            obj_ptr->set_name(divided_file[i+1]);

        if(s == "<followers>")
            inside_followers = true;

        if(s == "</followers>")
            inside_followers = false;

        if(s=="<post>"){
            post_ptr = new Post;
            if(divided_file[i+1]!="<body>")
                post_ptr->set_body(divided_file[i+1]);
        }

        if(s=="<body>")
            post_ptr->set_body(divided_file[i+1]);
        if(s=="<topic>")
            topics.push_back(divided_file[i+1]);


        if(s=="</post>"){
            posts.push_back(*post_ptr);
            delete post_ptr;
        }

        if(s=="</topics>"){
            post_ptr->set_topics(topics);
            topics.clear();
        }


        if(s=="</posts>") {
            obj_ptr->set_posts(posts);
            posts.clear();
        }

        if(s=="</user>"){
            g.add_user(obj_ptr->get_user_id(),*obj_ptr);
            delete obj_ptr;
        }
    }
    return g;
}